opentracing-java8/src/main/java/opentracing/propagation/SpanImpl.java

```diff
@@ -0,0 +1,115 @@
+/**
+ * Copyright 2016 The OpenTracing Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package opentracing.propagation;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import opentracing.Span;
+
+final class SpanImpl implements Span {
+
+    private final String operationName;
+    private final Optional<Span> parent;
+    private final Instant start;
+    private Duration duration;
+    final Map<String,String> traceState = new HashMap<>();
+    final Map<String,String> traceAttributes = new HashMap<>();
+    private final Map<String,Object> tags = new HashMap<>();
+    private final List<LogData> logs = new ArrayList<>();
+
+    public SpanImpl(String operationName, Optional<Span> parent) {
+        this.operationName = operationName;
+        this.parent = parent;
+        this.start = Instant.now();
+    }
+
+    SpanImpl(String operationName, Map<String,String> traceState) {
+        this(operationName, Optional.<Span>empty());
+        this.traceState.putAll(traceState);
+    }
+
+    @Override
+    public void finish() {
+        assert null == duration;
+        duration = Duration.between(start, Instant.now());
+    }
+
+    @Override
+    public Span setTag(String key, String value) {
+        tags.put(key, value);
+        return this;
+    }
+
+    @Override
+    public Span setTag(String key, boolean value) {
+        tags.put(key, value);
+        return this;
+    }
+
+    @Override
+    public Span setTag(String key, Number value) {
+        tags.put(key, value);
+        return this;
+    }
+
+    @Override
+    public Span setBaggage(String key, String value) {
+        traceAttributes.put(key, value);
+        return this;
+    }
+
+    @Override
+    public String getBaggage(String key) {
+        return traceAttributes.get(key);
+    }
+
+    @Override
+    public Span log(String message, /* @Nullable */ Object payload) {
+        Instant now = Instant.now();
```

Comment:
cool. hiding this allows use of the same interface regardless of java 8 vs java 7 (or even regardless of library)